sdks/go/pkg/beam/runners/dataflow/dataflowlib/metrics.go

```diff
@@ -0,0 +1,125 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dataflowlib translates a Beam pipeline model to the
+// Dataflow API job model, for submission to Google Cloud Dataflow.
+
+package dataflowlib
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/metrics"
+	df "google.golang.org/api/dataflow/v1b3"
+)
+
+// FromMetricUpdates extracts metrics from a slice of MetricUpdate objects and
+// groups them into counters, distributions and gauges.
+//
+// Dataflow currently only reports Counter and Distribution metrics to Cloud
+// Monitoring. Gauge metrics are not supported. The output metrics.Results will
+// not contain any gauges.
+func FromMetricUpdates(allMetrics []*df.MetricUpdate, job *df.Job) *metrics.Results {
+	ac, ad := groupByType(allMetrics, job, true)
+	cc, cd := groupByType(allMetrics, job, false)
+
+	return metrics.NewResults(metrics.MergeCounters(ac, cc), metrics.MergeDistributions(ad, cd), make([]metrics.GaugeResult, 0))
+}
+
+func groupByType(allMetrics []*df.MetricUpdate, job *df.Job, tentative bool) (
+	map[metrics.StepKey]int64,
+	map[metrics.StepKey]metrics.DistributionValue) {
+	counters := make(map[metrics.StepKey]int64)
+	distributions := make(map[metrics.StepKey]metrics.DistributionValue)
+
+	for _, metric := range allMetrics {
+		isTentative := metric.Name.Context["tentative"] == "true"
+		// Returns true when variables differ (exclusive or)
+		if (isTentative || tentative) && (!isTentative || !tentative) {
```

Comment:
If the goal here is to only provide either the tentative values OR the commited values, and not both, wouldn't
```
if isTentative != tentative {
  continue
}
```
be simpler?

Reply:
Right, I overcomplicated it. Thank you.

Reply:
No worries. I've been there myself.